firedrake/preconditioners/pmg.py

```diff
@@ -62,12 +62,42 @@ def coarsen_element(ele):
         degree = ele.degree()
         family = ele.family()
 
-        if family == "Discontinuous Galerkin" and degree == 0:
+        N = degree
+        try:
+            N, = set(N)
+        except TypeError:
+            pass
+        except ValueError:
+            raise NotImplementedError("Different degrees on TensorProductElement")
+
+        if family == "Discontinuous Galerkin" and N == 0:
             raise ValueError
-        elif degree == 1:
+        elif N == 1:
             raise ValueError
 
-        return ele.reconstruct(degree=degree // 2)
+        return PMGBase.reconstruct_degree(ele, N // 2)
+
+    @staticmethod
+    def reconstruct_degree(ele, N):
+        """
+        Reconstruct a given element, modifying its polynomial degree.
+
+        Must provide the same degree for tensor product elements.
+
+        :arg ele: a :class:`ufl.FiniteElement` to reconstruct.
+        :arg N: an integer degree.
+        """
+        if isinstance(ele, (VectorElement, TensorElement)):
+            cVe = [PMGBase.reconstruct_degree(K, N)
+                   for K in ele.sub_elements()]
+            return VectorElement(cVe[0], dim=len(cVe))
+        elif isinstance(ele, TensorProductElement):
+            (Ve, Qe) = ele.sub_elements()
```

Comment:
Do all `TensorProductElement`s only wrap two subelements?

Reply:
Good point, it should be extended to multiple subelements.